ML project/Customer_churn_prediction.py

```diff
@@ -0,0 +1,512 @@
+b_acc_lr = []
+acc_lr = []
+auc_lr = []
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    lr = LogisticRegression(C = 150, max_iter=600)
+    metrics = cross_val_score(lr, X_train, y_train, cv = 4, scoring=i)
+    if i == 'f1':
+        f1_lr.append(metrics)
+    elif i == 'balanced_accuracy':
+        b_acc_lr.append(metrics)
+    elif i == 'roc_auc':
+        auc_lr.append(metrics)
+    else:
+        acc_lr.append(metrics)
+
+
+# In[35]:
+
+
+plt.bar(['F1','Balanced accuracy','AUC','Accuracy'],[np.mean(f1_lr), np.mean(b_acc_lr),np.mean(auc_lr), np.mean(acc_lr)])
+plt.yticks([0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1])
+plt.title('LogR Metric score (average)');
+
+
+# In[36]:
+
+
+lr = LogisticRegression(C = 150, max_iter=600)
+lr.fit(X_train,y_train)
+preds_lr = lr.predict(X_test)
+preds_prob_lr = lr.predict_proba(X_test)[:,1]
+fpr_lr, tpr_lr, tresholds_lr = roc_curve(y_test,preds_prob_lr)
+prec_lr, recall_lr, tresh_lr = precision_recall_curve(y_test,preds_prob_lr)
+auc_score_lr = roc_auc_score(y_test,preds_prob_lr)
+
+
+# ## RandomForest
+
+# In[37]:
+
+
+from sklearn.ensemble import RandomForestClassifier
+rf = RandomForestClassifier()
+
+
+# In[38]:
+
+
+params = {'n_estimators' : [300,400,500],
+          'max_depth': [15,20,30],
+          'min_samples_leaf':[2,3]}
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rcv = RandomizedSearchCV(rf, param_distributions = params, cv=3,scoring = i,random_state=123)
+    rcv.fit(X_train,y_train)
+    print('Metrics: {}\nBest params: {}\nBest score: {}'.format(i,rcv.best_params_,rcv.best_score_))
+
+
+# In[39]:
+
+
+f1_rf = []
+b_acc_rf = []
+acc_rf = []
+auc_rf = []
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rf = RandomForestClassifier(n_estimators = 400, min_samples_leaf = 2, max_depth = 20)
+    metrics = cross_val_score(rf, X_train, y_train, cv = 10, scoring=i)
+    if i == 'f1':
+        f1_rf.append(metrics)
+    elif i == 'balanced_accuracy':
+        b_acc_rf.append(metrics)
+    elif i == 'roc_auc':
+        auc_rf.append(metrics)
+    else:
+        acc_rf.append(metrics)
+
+
+# In[40]:
+
+
+plt.bar(['F1','Balanced accuracy','AUC','Accuracy'],[np.mean(f1_rf), np.mean(b_acc_rf), np.mean(auc_rf),np.mean(acc_rf)])
+plt.yticks([0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1])
+plt.title('RF Metric score (average)');
+
+
+# In[41]:
+
+
+rf = RandomForestClassifier(n_estimators = 400, min_samples_leaf = 2, max_depth = 20)
+rf.fit(X_train,y_train)
+preds_rf = rf.predict(X_test)
+preds_prob_rf = rf.predict_proba(X_test)[:,1]
+fpr_rf, tpr_rf, tresholds_rf = roc_curve(y_test,preds_prob_rf)
+prec_rf, recall_rf, tresh_rf = precision_recall_curve(y_test,preds_prob_rf)
+auc_score_rf = roc_auc_score(y_test,preds_prob_rf)
+
+
+# In[42]:
+
+
+fi_df = pd.DataFrame(rf.feature_importances_, index = df_scaled.drop('churn',axis=1).columns, columns = ['feature_importance'])
+fi_df[fi_df.feature_importance>0.03].sort_values('feature_importance').plot(kind='barh');
```

Comment:
@Ettore1991 how would you do feature importance with Logistic Regression and Decision Tree?